scipy/_lib/_bunch.py

```diff
@@ -147,8 +147,11 @@ def __init__(self, {arg_list}, **extra_fields):
         self.__dict__[key] = val
 
 def __setattr__(self, key, val):
-    raise AttributeError("can't set attribute %r of class %r"
-                         % (key, self.__class__.__name__))
+    if key in {repr(field_names)}:
+        raise AttributeError("can't set attribute %r of class %r"
+                             % (key, self.__class__.__name__))
+    else:
+        self.__dict__[key] = val
```

Comment:
The point of this change is to allow setting attributes that were not part of the original tuple. Someone who deeply understands the bunch code should assess whether this is OK.

Reply:
I can move this stuff to a separate PR if it would help review.

Reply:
This change looks good to me. If we want to be more restrictive, we can just allow private attributes to be set (the one that start with `_`). Just a small thing though.

Reply:
I didn't really see the need to lock this down [originally](https://github.com/scipy/scipy/pull/12323#pullrequestreview-434215231). Sound ok to leave it open as-is?

Reply:
> why we are careful to ensure that the extra fields are not mutable and that none can be added?

I guess to be consistent with `namedtuple`.

> Sound ok to leave it open as-is?

Yup, I don't see any problem with it.